server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OnHeapSingleTreeBuilder.java

```diff
@@ -0,0 +1,204 @@
+        StarTreeField starTreeField,
+        DocValuesProducer docValuesProducer,
+        DocValuesConsumer docValuesConsumer,
+        SegmentWriteState segmentWriteState,
+        MapperService mapperService
+    ) throws IOException {
+        super(starTreeField, docValuesProducer, docValuesConsumer, segmentWriteState, mapperService);
+    }
+
+    @Override
+    public void appendStarTreeDocument(StarTreeDocument starTreeDocument) throws IOException {
+        starTreeDocuments.add(starTreeDocument);
+    }
+
+    @Override
+    public StarTreeDocument getStarTreeDocument(int docId) throws IOException {
+        return starTreeDocuments.get(docId);
+    }
+
+    @Override
+    public List<StarTreeDocument> getStarTreeDocuments() throws IOException {
+        return starTreeDocuments;
+    }
+
+    // TODO: should this be just long?
+    @Override
+    public long getDimensionValue(int docId, int dimensionId) throws IOException {
+        return starTreeDocuments.get(docId).dimensions[dimensionId];
+    }
+
+    // Handles star-tree rebuilds during merges :)
+    // @Override
+    // public void build(List<StarTreeValues> starTreeValues) throws IOException {
+    // TODO: This will be handled during off heap implementation for merges
+    // }
+
+    @Override
+    public Iterator<StarTreeDocument> processSegmentStarTreeDocuments(int numDocs) throws IOException {
+        StarTreeDocument[] starTreeDocuments = new StarTreeDocument[numDocs];
+        for (int i = 0; i < numDocs; i++) {
+            starTreeDocuments[i] = getNextSegmentStarTreeDocument();
+        }
+        return processStarTreeDocuments(starTreeDocuments);
+    }
+
+    /**
+     * Sort, aggregates and merges the star-tree documents
+     * @param starTreeDocuments star-tree documents
+     * @return iterator for star-tree documents
+     * @throws IOException throws when unable to sort, merge and aggregate star-tree documents
+     */
+    public Iterator<StarTreeDocument> processStarTreeDocuments(StarTreeDocument[] starTreeDocuments) throws IOException {
+
+        // sort the documents
+        Arrays.sort(starTreeDocuments, (o1, o2) -> {
+            for (int i = 0; i < numDimensions; i++) {
+                if (o1.dimensions[i] != o2.dimensions[i]) {
+                    return Math.toIntExact(o1.dimensions[i] - o2.dimensions[i]);
+                }
+            }
+            return 0;
+        });
+
+        // merge the documents
+        return mergeStarTreeDocuments(starTreeDocuments);
+    }
+
+    /**
+     * Merges the star-tree documents
+     * @param starTreeDocuments star-tree documents
+     * @return iterator to aggregate star-tree documents
+     */
+    private Iterator<StarTreeDocument> mergeStarTreeDocuments(StarTreeDocument[] starTreeDocuments) {
+        return new Iterator<>() {
+            boolean hasNext = true;
+            StarTreeDocument currentStarTreeDocument = starTreeDocuments[0];
+            int docId = 1;
```

Comment:
just checking, does docId start from 0 or 1 ?

Reply:
it would start from 1, as the 0th document is fetched on line 117.